File: src/CombinationSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by kaijiezhou on 7/9/15.
 */
public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<Integer> result=new ArrayList<>();
        List<List<Integer>> results=new ArrayList<>();
        findSum(results,result, candidates, target,0);
        return results;

    }

    //Function using DFS
    private void findSum(List<List<Integer>> results, List<Integer> result, int[] candidates, int target, int startIndex){
        if(target==0){
            /*We must create a new array,
              otherwise, it'll be changed by the following operation.
              The param "result" here is a pointer to a single array
             */
            results.add(new ArrayList(result));
        }else if(target>0){
            for(int i=startIndex;i<candidates.length;i++){
                result.add(candidates[i]);
                findSum(results, result, candidates, target-candidates[i],i);
                result.remove(result.size()-1);
            }
        }

    }
}

File: python/combination_sum.py
__author__ = 'kaijiezhou'
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        result=[]
        results=[]
        self.helper(results, result, candidates, target, 0)
        return results

    def helper(self, results, result, candidates, target, pos):
        if target==0:
            results.append(list(result))
        elif target>0:
            for i in xrange(pos, len(candidates)):
                result.append(candidates[i])
                self.helper(results, result, candidates, target-candidates[i], i)
                #remove from list by index
                result.pop(len(result)-1)


